src/mon/NVMeofGwSerialize.h

```diff
@@ -439,27 +527,53 @@ inline void decode(
     BeaconSubsystems   subsystems;
     decode(subsystems, bl);
     gw_created.subsystems = subsystems;
-    decode(size, bl);
-    for (uint32_t i=0; i<size; i++) {
-      uint32_t blklist_key;
-      uint32_t osd_epoch;
-      uint32_t is_failover;
-      decode(blklist_key, bl);
-      decode(osd_epoch,   bl);
-      decode(is_failover, bl);
-      Blocklist_data blst((epoch_t)osd_epoch, (bool)is_failover);
-
-      gw_created.blocklist_data[blklist_key] = blst;
+
+    if (struct_v >= 2) {
+      decode(size, bl);
+      for (uint32_t i=0; i<size; i++) {
+        uint32_t blklist_key;
+        uint32_t osd_epoch;
+        uint32_t is_failover;
+        decode(blklist_key, bl);
+        decode(osd_epoch,   bl);
+        decode(is_failover, bl);
+        Blocklist_data blst((epoch_t)osd_epoch, (bool)is_failover);
+        gw_created.blocklist_data[blklist_key] = blst;
+      }
+    } else {
+      for (uint32_t i=0; i<MAX_SUPPORTED_ANA_GROUPS; i++) {
+        uint32_t osd_epoch;
+        bool is_failover;
+        decode(osd_epoch,   bl);
+        dout(20) << "decode osd epoch  " << osd_epoch << dendl;
+        decode(is_failover, bl);
+        dout(20) << "decode is-failover  " << is_failover << dendl;
+        Blocklist_data blst((epoch_t)osd_epoch, (bool)is_failover);
+        // the same action as with "states"
+        gw_created.blocklist_data[i] = blst;
+      }
     }
     decode(gw_created.nonce_map, bl);
     gws[gw_name] = gw_created;
   }
+  if (struct_v == 1) {  //Fix allocations of states and blocklist_data
+    //since only after full loop on gws we know what states are relevant
+    for (auto &gw_it:gws) {
+      auto &state = gw_it.second;
+      for (uint32_t i=0; i<MAX_SUPPORTED_ANA_GROUPS; i++) {
```

Comment:
So, the difference between the two encodings is that the nvmeof-reef encoding always populates sm_state with 16 entries, regardless of which ana groups actually exist, right?  Rather than maintaining anas above (which will be an out of bounds array write if a group has an ana_grp_id >= 16) just iterate over gws here and build a std::set of the ids in gws.  Then, check that set for membership.  It's a bit less efficient, but this code path won't actually happen once everything is upgraded, so we want to make it as simple as possible.

Reply:
ok, done